### PR TITLE
Bulk edit improvements

### PR DESCRIPTION
**Pull request by szok**
Added:
- autofill
- back to list link
- hide checkbox in bulk edit mode
- permission to bulk edit fields

Connected to #1655

### CODE REVIEW

File: src/ralph/lib/permissions/models.py [hunk 1]
```diff
@@ -116,9 +115,19 @@ def has_access_to_field(cls, field_name, user, action='change'):
             cls._meta.model_name,
             field_name
         )
-        return user.has_perm(
+        perm = user.has_perm(
             '{}.{}'.format(cls._meta.app_label, perm_key)
         )
+        if action == 'view' and not perm:
```
Comment (mkurek) [inline src/ralph/lib/permissions/models.py:123]: add some comment why are you checking 'change' permission

File: src/ralph/static/src/js/fill-fields.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            results = matcher.exec(input_id);
+            sourceField = $('#' + input_id);
+        } else if (input_name !== '') {
+            results = matcher.exec(input_name);
+            sourceField = $('input[name="' + input_name + '"]');
+        }
+        pre = results[1];
+        number = results[2];
+        post = results[3];
+        for (var i = 0; i < MAX_ROWS; i++) {
```
Comment (mkurek) [inline src/ralph/static/src/js/fill-fields.js]: try to use some regex to get all inputs and operate on returned jquery "list" without loop

File: src/ralph/static/src/scss/_overrides.scss [hunk 3]
```diff
@@ -77,3 +77,15 @@ select {
         float: left;
     }
 }
+.result-list {
+    overflow-y: auto;
+}
+.bulk-edit-fill {
```
Comment (ar4s) [inline src/ralph/static/src/scss/_overrides.scss]: move to `./components/bulk_edit.scss` add some namespace (e.g. `.bulk-edit`) and include in `ralph.scss`

**Comment by vi4m** [comment]
Why we didn't use django admin actions framework ? It doesn't need custom template in filtering list 

https://docs.djangoproject.com/en/1.8/ref/contrib/admin/actions/

**Comment by szok** [comment]
We use. Bulk edit is added as an admin admin action. Additional parameters to the template allow us to add more features and hide checkbox list.

File: src/ralph/admin/templates/admin/widgets/autocomplete.html [hunk 4]
```diff
@@ -8,7 +8,7 @@
                 <i class="fa fa-times delete"></i>
             </div>
         </div>
-        <input class="input" type="text" placeholder="{% trans 'Search by' %} {{ searched_fields }}"/>
+        <input class="input no-fillable" type="text" placeholder="{% trans 'Search by' %} {{ searched_fields }}"/>
```
Comment (ghost) [inline src/ralph/admin/templates/admin/widgets/autocomplete.html]: Auto filling doesn't work on related fields? Is it other ticket?

Comment (szok) [inline src/ralph/admin/templates/admin/widgets/autocomplete.html]: Yes :)

File: src/ralph/admin/templates/admin/change_list.html [hunk 5]
```diff
@@ -89,24 +94,24 @@
...
-          {% if cl.has_filters %}
-            <div class="small-2 columns">
-              {% admin_actions %}
-              <div class="panel">
-                <h4>Filters</h4>
-                {% for spec in cl.filter_specs %}
-                  {% admin_list_filter cl spec %}
-                {% endfor %}
-              </div>
+        {% if not bulk_edit %}
```
Comment (ghost) [inline src/ralph/admin/templates/admin/change_list.html:87]: We should watch template files. This gonna soon look like ified and tagged soup (if already isn't, imho).

File: src/ralph/static/src/js/fill-fields.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    // We use absolute positioning for toolbar therefore body must be relative
+    $('body').css('position', 'relative');
+    // Disable autocomplete without cluttering html attributes
+    $('input').attr('autocomplete', 'off');
+
+    $('.result-list input[type=text]:not(.no-fillable), .result-list select').blur(function () {
+        $('#float_toolbar').hide();
+    });
+
+    var toggle_toolbar = function (input) {
```
Comment (ghost) [inline src/ralph/static/src/js/fill-fields.js]: This probably won't pass jshint.

Comment (mkurek) [inline src/ralph/static/src/js/fill-fields.js]: :+1:
we should run jshint on travis

Comment (ar4s) [inline src/ralph/static/src/js/fill-fields.js]: :+1:

Comment (szok) [inline src/ralph/static/src/js/fill-fields.js]: :+1:

Comment (quamilek) [inline src/ralph/static/src/js/fill-fields.js]: :+1:

Comment (andrzej-jankowski) [inline src/ralph/static/src/js/fill-fields.js]: :+1:

**Comment by ghost** [comment]
:ok:

**Comment by vi4m** [comment]
So please merge it if all comments resolved.

File: src/ralph/admin/templates/admin/change_list.html [hunk 7]
```diff
@@ -89,24 +94,24 @@
 {#             </div> #}
 {#           {% endif %} #}
 {#       </div> #}
```
Comment (mkurek) [inline src/ralph/admin/templates/admin/change_list.html]: is this commented block necessary?

File: src/ralph/lib/permissions/models.py [hunk 8]
```diff
@@ -116,9 +115,21 @@ def has_access_to_field(cls, field_name, user, action='change'):
             cls._meta.model_name,
             field_name
         )
-        return user.has_perm(
+        perm = user.has_perm(
             '{}.{}'.format(cls._meta.app_label, perm_key)
         )
+        # If the user does not have rights to view,
+        # but has the right to change it can view the field
```
Comment (mkurek) [inline src/ralph/lib/permissions/models.py]: he can view

File: src/ralph/admin/templates/admin/change_list.html (hunk 7 above)
Comment (szok) [inline src/ralph/admin/templates/admin/change_list.html]: No, I will remove

**Comment by mkurek** [comment]
:ok:

File: src/ralph/admin/mixins.py [hunk 9]
```diff
@@ -85,11 +85,24 @@ def changelist_view(self, request, extra_context=None):
         for view in self.list_views:
             views.append(view)
         extra_context['list_views'] = views
+        if self.get_actions(request) or self.list_filter:
+            extra_context['has_filters'] = True
+
+        extra_context['bulk_edit'] = request.GET.get(BULK_EDIT_VAR, False)
+        if extra_context['bulk_edit']:
+            extra_context['has_filters'] = False
```
Comment (ar4s) [inline src/ralph/admin/mixins.py:93]: template var `has_filters` is strong attached to visibility sidebar - maybe rename this var to `has_sidebar`? @szok what do you think?

File: src/ralph/admin/mixins.py [hunk 10]
```diff
@@ -85,11 +85,24 @@ def changelist_view(self, request, extra_context=None):
         for view in self.list_views:
             views.append(view)
         extra_context['list_views'] = views
+        if self.get_actions(request) or self.list_filter:
```
Comment (ar4s) [inline src/ralph/admin/mixins.py:88]: `extra_context['has_filters'] = bool(self.get_actions(request) or self.list_filter)`

**Comment by ar4s** [comment]
:ok:

**Pull request merged**